subprojects/dependency-management/src/main/java/org/gradle/internal/resolve/caching/FullAttributeContainerSerializer.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.internal.resolve.caching;
+
+import org.gradle.api.Named;
+import org.gradle.api.attributes.Attribute;
+import org.gradle.api.attributes.AttributeContainer;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.AttributeContainerSerializer;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.api.internal.attributes.ImmutableAttributesFactory;
+import org.gradle.api.internal.changedetection.state.CoercingStringValueSnapshot;
+import org.gradle.api.internal.model.NamedObjectInstantiator;
+import org.gradle.internal.serialize.Decoder;
+import org.gradle.internal.serialize.Encoder;
+
+import java.io.IOException;
+
+/**
+ * A lossy attribute container serializer. It's lossy because it doesn't preserve the attribute
+ * types: it will serialize the contents as strings, and read them as strings, only for reporting
+ * purposes.
+ */
+public class FullAttributeContainerSerializer implements AttributeContainerSerializer {
```

Comment:
:x: The javadocs are wrong, copy/pasted from the "lossy" version. I would also suggest to use a different name, like `DesugaringAttributeContainerSerializer` for this one (it accepts any attribute, and desugars them if needed) vs `DesugaredAttributeContainerSerializer` for the "lossy" one (it only accepts _already_ desugared attributes).

Reply:
Done